File: playService/assets/src/lobby/lobby.js
var snabbdom = require('snabbdom')
var patch = snabbdom.init([
    require('snabbdom/modules/eventlisteners').default,
    require('snabbdom/modules/props').default,
    require('snabbdom/modules/class').default,
    require('snabbdom/modules/style').default
])

var lobbyUi = require('./lobby-ui')
var lobbyService = require('./lobby-service')
var actions = require('./lobby-actions')

function Lobby(element, player) {
    var node = element
    const lobbyUpdates = lobbyService.updates()

    function renderState(state) {
        lobbyUpdates.onmessage = lobbyEventHandler(state)
        window.currentState = state
        node = patch(node, lobbyUi(state, actionHandler(state)))
    }

    function actionHandler(state) {
        return function(action) {
            const promise = actions(state, action)
            if (promise && promise.then)
                promise.then(renderState)
        }
    }

    function lobbyEventHandler(state){
        return function(event){
          state.games = JSON.parse(event.data)
          renderState(state)
        }
    }

    lobbyService.getAllGames().then(function (games) {
        renderState({player, games}, element)
    })

    return this
}

module.exports = Lobby

File: frontend/src/main/scala/roborace/frontend/lobby/Lobby.scala
package roborace.frontend.lobby

import gameEntities.GameResponse
import io.circe.generic.auto._
import io.circe.parser._
import org.scalajs.dom.raw.Element
import roborace.frontend.util.SnabbdomApp
import roborace.frontend.{LobbyFrontendState, Service}
import snabbdom.VNode

import scala.concurrent.ExecutionContext.Implicits.global

import scala.scalajs.js.|

class Lobby(container: Element) extends SnabbdomApp {

  var node: Element | VNode = container

  val eventSource = Service.lobbyUpdates()

  def renderState(state: LobbyFrontendState): Unit = {
    eventSource.onmessage = message => {
      val newGames = decode[Seq[GameResponse]](message.data.asInstanceOf[String]).right.get
      renderState(state.copy(games = newGames))
    }
    node = patch(node, LobbyUi.render(state).toVNode)

  }

  for {
    games     <- roborace.frontend.Service.getAllGames()
    scenarios <- roborace.frontend.Service.getAllScenarios()
  } renderState(LobbyFrontendState(games, scenarios))
}
